src/MultiMessageSender.sol

```diff
@@ -133,104 +148,165 @@ contract MultiMessageSender {
+            return;
+        }
 
+        uint256 i;
+        uint256 j;
+        uint256 k;
+        address[] memory merged = new address[](existings.length + _additions.length);
+        while (i < existings.length && j < _additions.length) {
+            address existing = existings[i];
+            address added = _additions[j];
+            if (existing < added) {
+                merged[k] = existing;
+                unchecked {
+                    ++i;
+                }
+            } else if (existing == added) {
+                revert Error.DUPLICATE_SENDER_ADAPTER();
+            } else {
+                merged[k] = added;
+                unchecked {
+                    ++j;
+                }
+            }
+            unchecked {
+                ++k;
+            }
+        }
+        while (i < existings.length) {
+            merged[k] = existings[i];
             unchecked {
                 ++i;
+                ++k;
+            }
+        }
+        while (j < _additions.length) {
+            merged[k] = _additions[j];
+            unchecked {
+                ++j;
+                ++k;
             }
         }
+
+        senderAdapters = merged;
+        _logSenderAdapterUpdates(_additions, true);
     }
 
     /// @notice Remove bridge sender adapters
-    /// @param _adapters is the adapter address to remove
-    function removeSenderAdapters(address[] calldata _adapters) external onlyOwner {
-        for (uint256 i; i < _adapters.length;) {
-            _removeSenderAdapter(_adapters[i]);
+    /// @param _removals are the adapter addresses to remove
+    function removeSenderAdapters(address[] calldata _removals) external onlyOwner {
+        _checkAdaptersOrder(_removals);
 
-            unchecked {
-                ++i;
-            }
-        }
+        address[] memory existings = senderAdapters;
+        address[] memory filtered = _filterAdapters(existings, _removals);
+
+        senderAdapters = filtered;
+        _logSenderAdapterUpdates(_removals, false);
     }
 
     /*/////////////////////////////////////////////////////////////////
-                            EXTERNAL VIEW FUNCTIONS
+                            PRIVATE/INTERNAL FUNCTIONS
     ////////////////////////////////////////////////////////////////*/
 
-    /// @notice A helper function for estimating total required message fee by all available message bridges.
-    function estimateTotalMessageFee(
-        uint256 _dstChainId,
-        address _multiMessageReceiver,
-        address _target,
-        bytes calldata _callData,
-        uint256 _nativeValue
-    ) public view returns (uint256 totalFee) {
-        MessageLibrary.Message memory message =
-            MessageLibrary.Message(block.chainid, _dstChainId, _target, nonce, _callData, _nativeValue, 0);
-        bytes memory data;
+    function _remoteCall(RemoteCallArgs memory _args) private {
+        (address mmaReceiver, address[] memory adapters) = _checkRemoteCallArgs(
```

Comment:
we might also need to validate that the total of `args.fees` is less than or equal to `msg.value`

Reply:
Done.